source/server/config/network/http_connection_manager.cc

```diff
@@ -67,6 +67,22 @@ NetworkFilterFactoryCb HttpConnectionManagerFilterConfigFactory::createFilterFac
   };
 }
 
+} // namespace
+
+NetworkFilterFactoryCb HttpConnectionManagerFilterConfigFactory::createFilterFactory(
+    const Json::Object& json_http_connection_manager, FactoryContext& context) {
+  envoy::api::v2::filter::HttpConnectionManager http_connection_manager;
+  Config::FilterJson::translateHttpConnectionManager(json_http_connection_manager,
+                                                     http_connection_manager);
+  return createHttpConnectionManagerFilterFactory(http_connection_manager, context);
+}
+
+NetworkFilterFactoryCb HttpConnectionManagerFilterConfigFactory::createFilterFactoryFromProto(
+    const Protobuf::Message& config, FactoryContext& context) {
+  return createHttpConnectionManagerFilterFactory(
+      dynamic_cast<const envoy::api::v2::filter::HttpConnectionManager&>(config), context);
```

Comment:
The type doesn't appear in the config at all, all we have is a `google.protobuf.Struct` knowledge when we encounter an opaque config. We then convert this `Struct` to JSON and then load it back into a `Protobuf::Message` (superclass of all concrete proto types), by invoking `Protobuf::util::JsonStringToMessage(json, &message)`. This function makes use of the type information in the supplied `message` to figure out how to do the parsing of the JSON.

This is why we need `createConfigProto()`, it's essentially providing a JSON->message conversion factory via the proto message it returns. If you can think of a cleaner way to do this, I'm open to that as well (we could literally renamed to `createConfigProtoFactory` and give it some more abstract interface if you prefer).